evaluation/numpy/segmentation_and_tracking_quality.py

```diff
@@ -0,0 +1,294 @@
+# coding=utf-8
+# Copyright 2021 The Deeplab2 Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of the Segmentation and Tracking Quality (STQ) metric."""
+
+import collections
+from typing import Mapping, MutableMapping, Sequence, Text, Any
+import numpy as np
+
+_EPSILON = 1e-15
+
+
+def _update_dict_stats(stat_dict: MutableMapping[int, np.ndarray],
+                       id_array: np.ndarray):
+  """Updates a given dict with corresponding counts."""
+  ids, counts = np.unique(id_array, return_counts=True)
+  for idx, count in zip(ids, counts):
+    if idx in stat_dict:
+      stat_dict[idx] += count
+    else:
+      stat_dict[idx] = count
+
+
+class STQuality(object):
+  """Metric class for the Segmentation and Tracking Quality (STQ).
+
+  Please see the following paper for more details about the metric:
+
+  "STEP: Segmenting and Tracking Every Pixel", Weber et al., arXiv:2102.11859, 
+  2021.
+
+
+  The metric computes the geometric mean of two terms.
+  - Association Quality: This term measures the quality of the track ID
+      assignment for `thing` classes. It is formulated as a weighted IoU
+      measure.
+  - Segmentation Quality: This term measures the semantic segmentation quality.
+      The standard class IoU measure is used for this.
+
+  Example usage:
+
+  stq_obj = segmentation_tracking_quality.STQuality(num_classes, things_list,
+    ignore_label, label_bit_shift, offset)
+  stq_obj.update_state(y_true_1, y_pred_1)
+  stq_obj.update_state(y_true_2, y_pred_2)
+  ...
+  result = stq_obj.result()
+  """
+
+  def __init__(self,
+               num_classes: int,
+               things_list: Sequence[int],
+               ignore_label: int,
+               label_bit_shift: int,
+               offset: int
+               ):
+    """Initialization of the STQ metric.
+
+    Args:
+      num_classes: Number of classes in the dataset as an integer.
+      things_list: A sequence of class ids that belong to `things`.
+      ignore_label: The class id to be ignored in evaluation as an integer or
+        integer tensor.
+      label_bit_shift: The number of bits the class label is shifted as an
```

Comment:
Even though label_bit_shift yields faster computation speed, I am worried that it is not compatible with our current setting in other places.
More specifically, we have panoptic_label_divisor = 1000 for KITTI-STEP. If users train a model with that setting, how could they use this script to evaluate their exported results?

Reply:
Please note that this encoding ( * label_divisor) is only (!) used internally in DeepLab2. Following our guides, e.g., for creating KITTI-STEP, the ground-truth is in 3 channel format (RGB) with R containing the semantic class. Furthermore, everything that is saved to disk is again 3 channel output (or at least not single-channel). 

The test servers expect 3 channel files. (Explained clearly at each test server.) Hence, I don't think it's a problem. For Deeplab2, the TF version must be used anyway.

What do you think?

Reply:
I see your point. You are right about it. Sorry for the false alarm. Thanks for the clarification, Mark!